app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java

```diff
@@ -0,0 +1,251 @@
+import android.widget.ImageView;
+
+import com.google.zxing.integration.android.IntentIntegrator;
+import com.google.zxing.integration.android.IntentResult;
+import com.nervousfish.nervousfish.ConstantKeywords;
+import com.nervousfish.nervousfish.R;
+import com.nervousfish.nervousfish.data_objects.Contact;
+import com.nervousfish.nervousfish.data_objects.IKey;
+import com.nervousfish.nervousfish.data_objects.KeyPair;
+import com.nervousfish.nervousfish.modules.cryptography.IKeyGenerator;
+import com.nervousfish.nervousfish.modules.database.IDatabase;
+import com.nervousfish.nervousfish.modules.qr.QRGenerator;
+import com.nervousfish.nervousfish.service_locator.IServiceLocator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import cn.pedant.SweetAlert.SweetAlertDialog;
+
+
+/**
+ * The activity that exchanges public keys through QR codes.
+ */
+@SuppressWarnings("checkstyle:ClassFanOutComplexity")
+//  1)  This warning is because the class relies on too many external classes, which can't really be avoided
+public class QRExchangeKeyActivity extends AppCompatActivity {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger("QRExchangeKeyActivity");
+
+    private IServiceLocator serviceLocator;
+
+    private AlertDialog lastDialog;
+
+
+    private IKey publicKey;
+
+    /**
+     * Creates the new activity, should only be called by Android
+     *
+     * @param savedInstanceState Don't touch this
+     */
+    @Override
+    protected void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        this.setContentView(R.layout.activity_qrexchange);
+
+        final Intent intent = getIntent();
+        this.serviceLocator = (IServiceLocator) intent.getSerializableExtra(ConstantKeywords.SERVICE_LOCATOR);
+
+
+        //TODO: Get the user's generated public key from the database
+        final IKeyGenerator keyGenerator = serviceLocator.getKeyGenerator();
+        final KeyPair pair = keyGenerator.generateRSAKeyPair("test");
+        publicKey = pair.getPublicKey();
+
+
+        final Button scanButton = (Button) findViewById(R.id.scanbutton);
+        scanButton.setOnClickListener(new View.OnClickListener() {
+
+            @Override
+            public void onClick(final View v) {
+                LOGGER.info("Started scanning QR code");
+                final IntentIntegrator integrator = new IntentIntegrator(QRExchangeKeyActivity.this);
+                integrator.initiateScan();
+
+            }
+        });
+
+        final Button generateButton = (Button) findViewById(R.id.generateQRbutton);
+        generateButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(final View v) {
+                LOGGER.info("Started generating QR code");
+                final String space = " ";
+                final Bitmap qrCode = QRGenerator.encode(publicKey.getType() + space + publicKey.getName()
```

Comment:
You could giva a try to the bzip lib. bzip usually has better compression ratios at speed expense. Some implementations below:

http://www.kohsuke.org/bzip2//

http://code.google.com/p/jbzip2/

http://www.java2s.com/Code/Java/File-Input-Output/BZip2format.htm

There is also an implementation of bzip on the apache commons compress lib:

http://commons.apache.org/compress/